### PR TITLE
feat(sink): Support Sinking Enum16 Ids to Clickhouse

### PR DESCRIPTION
**Pull request by HurricanKai**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This is an attempt at solving #14658 it simply allows the Clickhouse connector to sink Int16 / smallint as Enum16s to clickhouse. This "just works" although it's on the user to make sure all data provided is valid.

I've also tested Enum8, but presumably this has slightly different encoding, so causes CH to reject the row.

## Checklist

- [ ] I have written necessary rustdoc comments - I believe this is unecessary?
- [x] I have added necessary unit tests and integration tests - Could add E2E test but not sure, let me know what you want here. The change is pretty minimal.
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)
TODO Will do this if this change is generally received positively! (Just a tiny change in the Data Type Mapping in the docs).

## Release note

Allow sinking Enum16 values to Clickhouse

### CODE REVIEW

File: src/connector/src/sink/clickhouse.rs
Comment (hzxa21) [inline src/connector/src/sink/clickhouse.rs:263]: Since `Enum16` is sepecial, can you add some documentations here to explain the behavior and include the issue link?

**Review by hzxa21 — COMMENTED** [review]
Thanks for the contribution. Generally LGTM.

Would you mind adding a e2e test in https://github.com/risingwavelabs/risingwave/blob/9f9294e5957a9d0d89c3e709b8672c6dc5b72e9b/e2e_test/sink/clickhouse_sink.slt ? You need to modify the CK table schema and the result check [here](https://github.com/risingwavelabs/risingwave/blob/9f9294e5957a9d0d89c3e709b8672c6dc5b72e9b/ci/scripts/e2e-clickhouse-sink-test.sh#L34) accordingly as well.

**Comment by HurricanKai** [comment]
Added E2E tests & a comment. I couldn't get `./risedev d` to actually run, it just hangs lld for me. Seems like the E2E tests are covered by CI so :crossed_fingers: that passes

File: e2e_test/sink/clickhouse_sink.slt
Comment (xxhZs) [inline e2e_test/sink/clickhouse_sink.slt:8]: m6.v4 -> mv6.v4

**Review by xxhZs — COMMENTED** [review]
e2e ci is error, please fix and try again

File: ci/scripts/e2e-clickhouse-sink-test.sh
Comment (hzxa21) [inline ci/scripts/e2e-clickhouse-sink-test.sh:46]: $4 should be `B`

File: ci/scripts/e2e-clickhouse-sink-test.sh
Comment (hzxa21) [inline ci/scripts/e2e-clickhouse-sink-test.sh:47]: $4 should be `A`

**Review by hzxa21 — COMMENTED** [review]
It seems that the ordering of the result check in `e2e-clickhouse-sink-test.sh` doesn't match the ordering of the INSERT values. That is why you mistakenly put the wrong expected value in `$4`. Feel free to reorder the check.

**Comment by HurricanKai** [comment]
I was away so just got to updating it. Everything seems to be working now! Thanks for the help :smile:

**Review by stdrc — COMMENTED** [review]
I am personally a bit concerned about this PR. It seems rely heavily on the internal representation/encoding of `Enum` in ClickHouse, which may be subject to change. As you said, `Enum8` does not work as expected because it's encoding may differ from `Enum16`, but we cannot find any explanation on the difference in ClickHouse docs. So I don't think this is type safe.

Would it be better if we can support `Enum` by enhancing `ClickHouseField`? I see the crate `clickhouse` supports `Enum` type while I am not sure if it's possible to utilize it in RisingWave.

**Comment by HurricanKai** [comment]
Clickhouse will of course reject any rows that are not possible, when enum values don't exist, imo it's reasonable to ask a user to handle this.

Otherwise, it is type safe. Supporting enum8 is a matter of wiring through the Info that this is an enum and should be written as 8
bits instead. 
I didn't find any enum support in the crate, I might've overlooked it. I did ask on the ClickHouse slack, and learned that when writing rows using any of the efficient formats, writing as an integer directly is the only option, and converting between string and integers is the job of the user.
Also the "layout" of enums is well defined. Even if you omit specifying numbers in the definition it is [defined](https://clickhouse.com/docs/en/sql-reference/data-types/enum) that it will be numbered sequentially starting from 1 or the value assigned to the first entry.

**Comment by stdrc** [comment]
> I didn't find any enum support in the crate, I might've overlooked it.

On the homepage of the crate on docs.rs:

截屏2024-02-19 13 29 54

But it seems not easy to utilized in this PR's situation.

> I did ask on the ClickHouse slack, and learned that when writing rows using any of the efficient formats, writing as an integer directly is the only option, and converting between string and integers is the job of the user.

Then looks good to me. May wait for code owner's approval.

**Comment by HurricanKai** [comment]
I see, thank you! I think that example only works because `#[repr(u8)]` essentially makes the enum transparently a `u8` in rusts compiler, and it'll be automatically converted to u8 when passed to the crate. It's a nice convenience feature when using the crate directly, but as you say, can't make use of that here as the enums are dynamically created by the user.

**Review by xxhZs — APPROVED** [review]
LGTM! Thanks

**Pull request merged**
